Restrict fullscreen exit on input focus to text-entry elements only

nullhook opened the pull request:
#### e34179969750a466a470d7da21a5e09bb680eec3
<pre>
Restrict fullscreen exit on input focus to text-entry elements only
<a href="https://bugs.webkit.org/show_bug.cgi?id=292220">https://bugs.webkit.org/show_bug.cgi?id=292220</a>
<a href="https://rdar.apple.com/136726993">rdar://136726993</a>

Reviewed by Brent Fulgham.

Previously, we exit fullscreen when any form control element
received focus. This behavior was overly broad.

Non-text input types like sliders, checkboxes, and radios
will no longer trigger fullscreen exit

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::isTextFormControlOrEditableContent):

Canonical link: <a href="https://commits.webkit.org/294284@main">https://commits.webkit.org/294284@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d546c424bdfde9a95fc10a501524f778661888ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101234 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20896 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11199 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106382 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51861 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29390 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77096 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34132 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104241 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16347 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91441 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57443 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16163 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51209 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86048 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9527 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108738 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28362 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20870 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86073 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85623 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30350 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8067 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22461 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16492 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28292 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33563 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28104 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31424 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29662 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7395,7 +7395,10 @@ bool WebPage::shouldDispatchUpdateAfterFocusingElement(const Element& element) c
 static bool isTextFormControlOrEditableContent(const WebCore::Element& element)
 {
-    return is<HTMLTextFormControlElement>(element) || element.hasEditableStyle();
+    if (auto input = dynamicDowncast<const HTMLInputElement>(element))
```
```suggestion
    if (RefPtr input = dynamicDowncast<const HTMLInputElement>(element))
```

brentfulgham replied:
+1 to @aprotyas 's suggestion

nullhook replied:
changed

nullhook commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7395,7 +7395,10 @@ bool WebPage::shouldDispatchUpdateAfterFocusingElement(const Element& element) c
 static bool isTextFormControlOrEditableContent(const WebCore::Element& element)
 {
-    return is<HTMLTextFormControlElement>(element) || element.hasEditableStyle();
+    if (auto input = dynamicDowncast<const HTMLInputElement>(element))
```
we might have to handle textarea check separately?

aprotyas commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7395,7 +7395,10 @@ bool WebPage::shouldDispatchUpdateAfterFocusingElement(const Element& element) c
 
 static bool isTextFormControlOrEditableContent(const WebCore::Element& element)
```
Consider renaming this since that's not what the function returns anymore?

nullhook replied:
left this function as-is and added a new one.

brentfulgham commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -7421,8 +7435,8 @@ void WebPage::elementDidFocus(Element& element, const FocusOptions& options)
-        if (element.document().fullscreen().isFullscreen())
-            element.document().fullscreen().fullyExitFullscreen();
+    if (shouldExitFullscreenAfterFocusingElement(element))
+        element.document().fullscreen().fullyExitFullscreen();
```
I like this version much better!

brentfulgham reviewed (approved):
r=me

brentfulgham commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -231,6 +231,7 @@
 #include <WebCore/HTMLModelElement.h>
 #include <WebCore/HTMLPlugInElement.h>
 #include <WebCore/HTMLSelectElement.h>
+#include <WebCore/HTMLTextAreaElement.h>
```
Nit: This could be limited to the `#if PLATFORM(IOS_FAMILY)` block of includes, but I don't think that's important.

nullhook committed:
Restrict fullscreen exit on input focus to text-entry elements only
https://bugs.webkit.org/show_bug.cgi?id=292220
rdar://136726993

Reviewed by Brent Fulgham.

Previously, we exit fullscreen when any form control element
received focus. This behavior was overly broad.

Non-text input types like sliders, checkboxes, and radios
will no longer trigger fullscreen exit

* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::isTextFormControlOrEditableContent):

Canonical link: https://commits.webkit.org/294284@main
